crates/matrix-sdk/src/encryption/secret_storage.rs

```diff
@@ -0,0 +1,484 @@
+    /// - `m.cross_signing.user_signing`: The user-signing cross-signing key.
+    ///
+    /// If the `m.cross_signing.self_signing` key is successfully imported, it
+    /// is used to sign our own [`Device`], marking it as verified. This step is
+    /// establishes trust in your own device's identity.
+    ///
+    /// By invoking this method, you ensure that your device has access to
+    /// the necessary secrets for device and identity verification.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::Client;
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use ruma::events::secret::request::SecretName;
+    ///
+    /// let secret_store = client
+    ///     .encryption()
+    ///     .open_secret_store("It's a secret to everybody")
+    ///     .await?;
+    ///
+    /// secret_store.import_secrets().await?;
+    ///
+    /// let status = client
+    ///     .encryption()
+    ///     .cross_signing_status()
+    ///     .await
+    ///     .expect("We should be able to check out cross-signing status");
+    ///
+    /// println!("Cross-signing status {status:?}");
+    ///
+    /// # anyhow::Ok(()) };
+    /// ```
+    ///
+    /// [`Device`]: crate::encryption::identities::Device
+    #[instrument(fields(user_id, device_id, cross_signing_status))]
+    pub async fn import_secrets(&self) -> Result<()> {
+        let olm_machine = self.client.olm_machine().await;
+        let olm_machine = olm_machine.as_ref().ok_or(crate::Error::NoOlmMachine)?;
+
+        Span::current()
+            .record("user_id", display(olm_machine.user_id()))
+            .record("device_id", display(olm_machine.device_id()));
+
+        info!("Fetching the private cross-signing keys from the secret store");
+
+        // Get all our private cross-signing keys from the secret store.
+        let export = self.get_cross_signing_keys().await?;
+
+        info!(cross_signing_keys = ?export, "Received the cross signing keys from the server");
+
+        // We need to ensure that we have the public parts of the cross-signing keys,
+        // those are represented as the `OwnUserIdentity` struct. The public
+        // parts from the server are compared to the public parts re-derived from the
+        // private parts. We will only import the private parts of the cross-signing
+        // keys if they match to the public parts, otherwise we would risk
+        // importing some stale cross-signing keys leftover in the secret store.
+        let (request_id, request) = olm_machine.query_keys_for_users([olm_machine.user_id()]);
+        self.client.keys_query(&request_id, request.device_keys).await?;
+
+        // TODO: Import the backup key here as well if it exists and enable backups if
+        // the current backup version is trusted, or expose a different method for this?
```

Comment:
Is this `TODO` addressed, or do we have an issue to not forget about it?

Reply:
It's the thing I'm currently working on and everybody is anxious to get as soon as possible. There's very little chance that we're going to forget this one.